File: DeforaOS/Apps/Web/DaPortal/system/debug.php
<?php //$Id$
//Copyright (c) 2011 Pierre Pronchery <khorben@defora.org>
//This file is part of DaPortal
//
//DaPortal is free software; you can redistribute it and/or modify
//it under the terms of the GNU General Public License version 2 as
//published by the Free Software Foundation.
//
//DaPortal is distributed in the hope that it will be useful,
//but WITHOUT ANY WARRANTY; without even the implied warranty of
//MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//GNU General Public License for more details.
//
//You should have received a copy of the GNU General Public License
//along with DaPortal; if not, write to the Free Software
//Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA



$debug_messages = '';
//FIXME force a timezone or use a different function than gettimeofday()
$debug_time_before = gettimeofday();


function _debug()
{
	global $debug, $html, $debug_messages, $debug_time_before;

	if(!$debug || !$html)
		return;
	$debug_time_after = gettimeofday();
	$sec = $debug_time_after['sec'] - $debug_time_before['sec'];
	$usec = $debug_time_after['usec'] - $debug_time_before['usec'];
	if($usec < 0)
	{
		$usec = -$usec;
		$sec++;
	}
	_info('Page execution duration: '.$sec.'s and '.ceil($usec / 1000).'ms',
			FALSE);
	print('<div class="debug system">'."\n");
	print($debug_messages);
}


//PRE
//	$level is trusted
function _debug_message($level, $title, $message, $visible = FALSE)
{
	global $debug, $debug_messages, $html;

	if(!$html)
		return;
	if($visible)
		print('<div class="debug"><div class="visible '.$level.'">'
				.'<b>'.htmlspecialchars($title)
				.'</b><br/>'.htmlspecialchars($message)
				."</div></div>\n");
	if(!$debug)
		return;
	$debug_messages.="\t".'<div class="'.$level.'"><b>'.(ucfirst($level))
			.':</b> '.htmlspecialchars($message)."</div>\n";
}


function _error($message, $visible = TRUE)
{
	global $engine;

	$engine->log('LOG_ERR', $message);
	if($visible)
		_debug_message('error', 'Error', $message, $visible);
}


function _info($message, $visible = FALSE)
{
	global $engine;

	$engine->log('LOG_WARN', $message);
	if($visible)
		_debug_message('info', 'Information', $message, $visible);
}


function _warning($message, $visible = FALSE)
{
	global $engine;

	$engine->log('LOG_WARN', $message);
	if($visible)
		_debug_message('warning', 'Warning', $message, $visible);
}

?>

File: DeforaOS/Apps/Desktop/src/Locker/src/plugins/debug.c
/* $Id$ */
/* Copyright (c) 2011-2012 Pierre Pronchery <khorben@defora.org> */
/* This file is part of DeforaOS Desktop Locker */
/* This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 3 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>. */



#include <System.h>
#include <stdio.h>
#include "Locker.h"


/* Debug */
/* private */
/* types */
typedef struct _LockerPlugin
{
	LockerPluginHelper * helper;
} Debug;


/* prototypes */
/* plug-in */
static Debug * _debug_init(LockerPluginHelper * helper);
static void _debug_destroy(Debug * debug);
static void _debug_event(Debug * debug, LockerEvent event);


/* public */
/* variables */
/* plug-in */
LockerPluginDefinition plugin =
{
	"Debug",
	"applications-development",
	NULL,
	_debug_init,
	_debug_destroy,
	_debug_event
};


/* private */
/* functions */
/* debug_init */
static Debug * _debug_init(LockerPluginHelper * helper)
{
	Debug * debug;

	if((debug = object_new(sizeof(*debug))) == NULL)
		return NULL;
	debug->helper = helper;
	fprintf(stderr, "DEBUG: %s()\n", __func__);
	return debug;
}


/* debug_destroy */
static void _debug_destroy(Debug * debug)
{
	fprintf(stderr, "DEBUG: %s()\n", __func__);
	object_delete(debug);
}


/* debug_event */
static void _debug_event(Debug * debug, LockerEvent event)
{
	switch(event)
	{
		case LOCKER_EVENT_ACTIVATING:
			fprintf(stderr, "DEBUG: %s() ACTIVATING\n", __func__);
			return;
		case LOCKER_EVENT_LOCKING:
			fprintf(stderr, "DEBUG: %s() LOCKING\n", __func__);
			return;
		case LOCKER_EVENT_SUSPENDING:
			fprintf(stderr, "DEBUG: %s() SUSPENDING\n", __func__);
			return;
		case LOCKER_EVENT_UNLOCKING:
			fprintf(stderr, "DEBUG: %s() UNLOCKING\n", __func__);
			return;
	}
	fprintf(stderr, "DEBUG: %s() Unknown event (%u)\n", __func__, event);
}
